lib/grandpa/catch-up.go

```diff
@@ -0,0 +1,281 @@
+	}
+
+	if msg.SetID != c.grandpa.state.setID {
+		return fmt.Errorf("%w: received set id %d but have set id %d in state",
+			ErrSetIDMismatch, msg.SetID, c.grandpa.state.setID)
+	}
+
+	if msg.Round <= c.grandpa.state.round {
+		return fmt.Errorf("%w: received round %d but grandpa round in state is %d",
+			ErrInvalidCatchUpResponseRound, msg.Round, c.grandpa.state.round)
+	}
+
+	if c.bestResponse.Load().(*CatchUpResponse).Round >= msg.Round {
+		logger.Debug("ignoring catch up response, since we are already processing one with a higher round")
+	}
+
+	prevote, err := c.verifyPreVoteJustification(msg)
+	if err != nil {
+		return fmt.Errorf("cannot verify pre vote justification: %w", err)
+	}
+
+	if err = c.verifyPreCommitJustification(msg); err != nil {
+		return fmt.Errorf("cannot verify pre commit justification: %w", err)
+	}
+
+	if msg.Hash.IsEmpty() || msg.Number == 0 {
+		return ErrGHOSTlessCatchUp
+	}
+
+	if err = c.verifyCatchUpResponseCompletability(prevote, msg.Hash); err != nil {
+		return fmt.Errorf("cannot verify catch-up response completability: %w", err)
+	}
+
+	// set prevotes and precommits in db
+	if err = c.grandpa.grandpaState.SetPrevotes(msg.Round, msg.SetID, msg.PreVoteJustification); err != nil {
+		return fmt.Errorf("cannot set pre votes in grandpa state: %w", err)
+	}
+
+	if err = c.grandpa.grandpaState.SetPrecommits(msg.Round, msg.SetID, msg.PreCommitJustification); err != nil {
+		return fmt.Errorf("cannot set pre commits in grandpa state: %w", err)
+	}
+
+	// update state and signal to grandpa we are ready to initiate
+	head, err := c.grandpa.blockState.GetHeader(msg.Hash)
+	if err != nil {
+		logger.Debugf("failed to process catch up response for round %d, storing the catch up response to retry", msg.Round)
+		return fmt.Errorf("cannot get header from grandpa block state: %w", err)
+	}
+
+	c.grandpa.head = head
+	c.grandpa.state.round = msg.Round
+	if c.waitingOnResponse.Load().(bool) {
+		c.catchUpResponseCh <- msg
+	}
+
+	c.waitingOnResponse.Store(false)
+
+	// resetting both response and requests
+	c.bestResponse.Store(nil)
+	c.lock.Lock()
+	c.requestsSent = make(map[peer.ID]CatchUpRequest)
+	c.lock.Unlock()
+
+	logger.Debugf("caught up to round; unpaused service and grandpa state round is %d", c.grandpa.state.round)
+	return nil
+}
+
+// verifyCatchUpResponseCompletability verifies that the pre-commit block is a descendant of, or is, the pre-voted block
+func (c *catchUp) verifyCatchUpResponseCompletability(prevote, precommit common.Hash) error {
+	if prevote == precommit {
+		return nil
+	}
+
+	// check if the current block is a descendant of prevoted block
+	isDescendant, err := c.grandpa.blockState.IsDescendantOf(prevote, precommit)
+	if err != nil {
+		return fmt.Errorf("cannot find descendance between prevote and precommit: %w", err)
+	}
+
+	if !isDescendant {
+		return ErrCatchUpResponseNotCompletable
+	}
+
+	return nil
+}
+
+func (c *catchUp) verifyPreCommitJustification(msg *CatchUpResponse) error {
```

Comment:
this function already exists in `lib/grandpa/message_handler.go`, why not use it?